compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp

```diff
@@ -0,0 +1,252 @@
+
+    _expand_dims = _g.nodes()->create<luci::CircleExpandDims>();
+    _expand_dims->input(_input);
+    _expand_dims->axis(_axis);
+    _expand_dims->name("expand_dims");
+
+    _output->from(_expand_dims);
+  }
+
+protected:
+  luci::CircleConst *_axis = nullptr;
+
+protected:
+  luci::CircleExpandDims *_expand_dims = nullptr;
+};
+
+class SubstituteExpandDimsToReshapeTest : public ::testing::Test
+{
+public:
+  SubstituteExpandDimsToReshapeTest() = default;
+
+  void run_pass(void)
+  {
+    while (_shapeinf.run(_graph.g()) || _pass.run(_graph.g()))
+      ;
+  }
+
+protected:
+  ExpandDimsWithConstAxisGraph _graph; // *W*ith *C*onst axis
+  luci::SubstituteExpandDimsToReshapePass _pass;
+  luci::CircleShapeInferencePass _shapeinf;
+};
+
+class SubstituteExpandDimsToReshapeNegTest : public ::testing::Test
+{
+public:
+  SubstituteExpandDimsToReshapeNegTest() = default;
+
+  void run_pass(void)
+  {
+    while (_shapeinf.run(_graph.g()) || _pass.run(_graph.g()))
+      ;
+  }
+
+protected:
+  ExpandDimsWithInputAxisGraph _graph; // *W*ith *N*on *C*onst axis
+  luci::SubstituteExpandDimsToReshapePass _pass;
+  luci::CircleShapeInferencePass _shapeinf;
+};
+
+} // namespace
+
+TEST(SubstituteExpandDimsToReshapePassTest, name)
+{
+  luci::SubstituteExpandDimsToReshapePass pass;
+  auto const name = pass.name();
+  ASSERT_NE(nullptr, name);
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_1)
+{
+  _graph.init({2, 16}, {2, 1, 16}, 1);
+
+  run_pass();
+
+  auto reshape = dynamic_cast<luci::CircleReshape *>(_graph.output()->from());
+  auto expand_dims = dynamic_cast<luci::CircleExpandDims *>(_graph.output()->from());
+  ASSERT_NE(nullptr, reshape);
+  ASSERT_EQ(nullptr, expand_dims);
+  auto reshape_shape = loco::must_cast<luci::CircleConst *>(reshape->shape());
+  ASSERT_EQ(3, reshape_shape->size<loco::DataType::S32>());
+  ASSERT_EQ(2, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(16, reshape_shape->at<loco::DataType::S32>(2));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_M1)
+{
+  _graph.init({2, 3, 4}, {2, 3, 4, 1}, -1);
+
+  run_pass();
+
+  auto reshape = dynamic_cast<luci::CircleReshape *>(_graph.output()->from());
+  auto expand_dims = dynamic_cast<luci::CircleExpandDims *>(_graph.output()->from());
+  ASSERT_NE(nullptr, reshape);
+  ASSERT_EQ(nullptr, expand_dims);
+  auto reshape_shape = loco::must_cast<luci::CircleConst *>(reshape->shape());
+  ASSERT_EQ(4, reshape_shape->size<loco::DataType::S32>());
+  ASSERT_EQ(2, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(3, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(4, reshape_shape->at<loco::DataType::S32>(2));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(3));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_2)
+{
+  // NOTE Incorrect output shape is allowed
+  _graph.init({16, 3, 1}, {16, 3, 4, 1}, 2);
```

Comment:
@jinevening This is unexpected from my view. I tried to turn this into negative test by giving it an incorrect Output shape. This case, correct shape whill be ({16,3,1}, {16,3,1,1,}, 2). Do you have any idea why it runs without failure?

Reply:
That's because no one checks the shape is correct in the test code.

Reply:
@jinevening Even shape inference code doen't check?

Reply:
Shape inference only checks if inputs have invalid shapes fo inferring output shape. So, if inputs had valid shapes, the output shape would be inferred and overwritten even though the previous output shape was wrong.